src/main/java/uk/gov/digital/ho/hocs/audit/export/CustomExportService.java

```diff
@@ -65,23 +65,34 @@ public void customExport(HttpServletResponse response, String code, boolean conv
             if (convertHeader) {
                 substitutedHeaders = headerConverter.substitute(headers);
             }
-            List<Object[]> dataList = customExportDataConverter.convertData(auditRepository.getResultsFromView(exportViewDto.getCode()), exportViewDto.getFields());
 
             try (CSVPrinter printer = new CSVPrinter(outputWriter, CSVFormat.DEFAULT.withHeader(substitutedHeaders.toArray(new String[substitutedHeaders.size()])))) {
-
-                dataList.forEach((dataRow) -> {
-                    try {
-                        printer.printRecord(dataRow);
-                        outputWriter.flush();
-                    } catch (Exception e) {
-                        log.error("Unable to parse record for custom report, reason: {}, event: {}", e.getMessage(), value(LogEvent.EVENT, CSV_EXPORT_FAILURE));
-                    }
-                });
-                log.info("Export Custom Report '{}' to CSV Complete, event {}", exportViewDto.getCode(), value(EVENT, CSV_EXPORT_COMPETE));
+                retrieveAuditData(exportViewDto.getCode())
+                        .forEach(data -> {
+                            Object[] converted = customExportDataConverter.convertData(data, exportViewDto.getFields());
+
+                            if (converted == null) {
+                                log.warn("No data to print after converting data {}", data);
+                                return;
+                            }
+
+                            try {
+                                printer.printRecord(converted);
+                                outputWriter.flush();
+                            } catch (IOException e) {
+                                log.error("Unable to parse record for custom report, reason: {}, event: {}", e.getMessage(), value(LogEvent.EVENT, CSV_EXPORT_FAILURE));
+                            }
+                        });
             }
```

Comment:
Lovely implementation and example of streaming the records as they are extracted from the database and converted.